presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java

```diff
@@ -74,28 +65,39 @@ public BinaryFileSpiller(
         this.serde = requireNonNull(serde, "serde is null");
         this.executor = requireNonNull(executor, "executor is null");
         this.spillerStats = requireNonNull(spillerStats, "spillerStats is null");
-        this.localSpillContext = localSpillContext;
+        this.localSpillContext = requireNonNull(localSpillContext, "localSpillContext can not be null");
         try {
-            this.targetDirectory = Files.createTempDirectory(spillPath, "presto-spill");
+            targetFileName = Files.createTempFile(spillPath, "spill", ".bin");
         }
         catch (IOException e) {
-            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Failed to create spill directory", e);
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Failed to create spill file", e);
         }
     }
 
     @Override
     public ListenableFuture<?> spill(Iterator<Page> pageIterator)
     {
-        checkState(previousSpill.isDone());
-        Path spillPath = getPath(spillsCount++);
+        checkNoSpillInProgress();
+        spillInProgress = executor.submit(
+                () -> writePages(pageIterator));
+        return spillInProgress;
+    }
 
-        previousSpill = executor.submit(() -> writePages(pageIterator, spillPath));
-        return previousSpill;
+    private void checkNoSpillInProgress()
+    {
+        checkState(spillInProgress.isDone(), "spill in progress");
     }
 
-    private void writePages(Iterator<Page> pageIterator, Path spillPath)
+    @Override
+    public Iterator<Page> getSpilledPages()
```

Comment:
I think we should wait until we have more users of the spill code before we make a call on the interface, so I'm find leaving it for now.

What I would like to see is an interface that makes it impossible (difficult) to violate state machine, and makes usage of the spiller more obvious.  It isn't clear yet if we want to support multiple current background spills, or if we want to support async reads... so we should wait.